File: Engine/src/Modules/Stellar.Core/Data/Registry/IdentifierRegistry.cs
using System.Collections.Concurrent;
using Stellar.Kernel;
using Stellar.Kernel.Data.Registry;

namespace Stellar.Core.Data.Registry;

public class IdentifierRegistry
    : IRegistry<Identifier>
{
    private static readonly ConcurrentDictionary<Guid, Identifier> Identifiers = new();
    
    private static readonly Lazy<IdentifierRegistry> Registry = new(() => new IdentifierRegistry());
    public static IdentifierRegistry Instance => Registry.Value;

    public bool Exists(IIdentifier id)
    {
        return Identifiers.ContainsKey(id.UID);
    }

    public bool Register(Identifier obj)
    {
        return Identifiers.TryAdd(obj.UID, obj);
    }

    public Identifier? Get(Guid uid)
    {
        return Identifiers.GetValueOrDefault(uid);
    }

    public Identifier? Get(IIdentifier id)
    {
        return Get(id.UID);
    }

    public Identifier? Pop(Guid uid)
    {
        Identifiers.TryRemove(uid, out var obj);
        return obj;
    }

    public Identifier? Pop(IIdentifier id)
    {
        return Pop(id.UID);
    }

    public int Size => Identifiers.Count;
    public ICollection<IIdentifier> Keys => (ICollection<IIdentifier>)Identifiers.Values;
    public ICollection<Identifier> Values => Identifiers.Values;
}

File: Engine/src/Native/Stellar.Native.Core/Data/Registry/IdentifierRegistry.cpp
#include "pch.h"
#include "IdentifierRegistry.h"

namespace Stellar::Native::Core::Data::Registry
{
    bool IdentifierRegistry::Register(const Identifier& id)
    {
        return _identifiers.TryAdd(id, id);
    }

    std::optional<Identifier> IdentifierRegistry::Get(const Identifier& id) const
    {
        return _identifiers.TryGet(id);
    }

    bool IdentifierRegistry::Unregister(const Identifier& id, Identifier& outValue)
    {
        return _identifiers.TryRemove(id, outValue);
    }

    bool IdentifierRegistry::Contains(const Identifier& key) const
    {
        return _identifiers.Contains(key);
    }

    size_t IdentifierRegistry::size() const
    {
        return _identifiers.size();
    }

    std::generator<const Identifier&> IdentifierRegistry::Identifiers() const
    {
        return _identifiers.Keys();
    }
}
